### PR TITLE
cdi: add Digital Video Cartridge (VMPEG) emulation and ntsc support

### LINKED ISSUE #1170: Philipps CD-I [cdimono1] : no DIGITAL VIDEO cardtridge
Hello,

some CD-I like "7th guest" needs a DIGITAL VIDEO Cardtridge . 
You can not play this game without this extension card.
It is planned?

LGDN
Mame 0.176

### LINKED ISSUE #14018: CD-i: NTSC Support
### MAME version

0.278

### System information

N/A

### INI configuration details

```INI

```

### Emulated system/software

CD-i

### Incorrect behaviour

Image
Some NTSC games, such as `The Adventures of the Spaceship Beagle` are coded incorrectly by the developers and do not check if the game is being displayed on a PAL TV. Since they don't compensate for the screen size, the pixels are skewed.

Tampering with the PAL flag in the MCD212 chip code will allow the output to look correct, but there is currently no way to launch an NTSC version of the CD-i. The mono-II NTSC console currently doesn't boot.

### Expected behaviour

A working CD-i launch mode which allows for NTSC screensize.

### Steps to reproduce

1. Start the game Adventures of the Spaceship Beagle with the cdimono1 bios.
2. When the game starts, you will instantly see the opening credits are illegible.

3. You can manually disable the left and right width buffers in the MCD212 code, and you will see the output is now legible. If the MCD212 chip to receive the NTSC region flag, you can see that the code would produce the correct result.

### Additional details

_No response_

### PR DESCRIPTION
**Pull request by aderumier**
cdi: add Digital Video Cartridge (VMPEG) emulation, and an NTSC Mono-I (cdimono1n)

This adds emulation of the CD-i Digital Video Cartridge (the "VMPEG" MPEG-1
cartridge) to the Mono-I, so Digital Video titles now play: MPEG video (FMV)
and MPEG audio (FMA). It also adds an NTSC variant of the Mono-I, so NTSC-only
discs run.

Updated after review: the MPEG-1 video decoding now uses the `mpeg_video`
backend from #16053 instead of a bundled library, the audio uses MAME's
existing `mpeg_audio`, and the cartridge is split into the two chips a VMPEG
board carries, so either can be reused on its own.

**This series depends on #16053** for `src/devices/video/mpeg_video`, which is
included here as its own commit so this builds on its own; drop that commit
once #16053 is merged. It also carries @Vincent-Halver's scc68070 fix from
#15978, which the DVC needs and which is the same fix I had made
independently.

## Changes

**scc68070: Add Matching Memory Pairs** (@Vincent-Halver, from #15978).
Only channel 0's Memory Transfer Counter was decoded: writes to channel 1's
were discarded and reads returned zero, so every channel 1 transfer had a
length of zero. The CDIC uses channel 0; the DVC is the first user of
channel 1.

**video/mpeg_video: Expose what the CD-i DVC needs.**
Additions to the decoder from #16053, none of which change what `decode()`
returns: the sequence and group flags, the picture rate code and the GOP time
code the DVC status registers report; `complete_picture()`, so a picture ends
on its last macroblock when no more input is coming; and `reset_input()`, to
resume a stream mid sequence. Plus two fixes for a stream that ends right
after a picture, where the parser waited for bits that never arrive.

**sound/mpeg_audio: fix joint stereo band decoding.**
The loop reading the bands above the joint stereo bound tested the wrong
limit, so it never ran: those bands stayed zero and their bits were left in
the stream, and the rest of the frame decoded from the wrong bit position.
Plain stereo hid it. CD-i digital video audio alternates stereo and joint
stereo from frame to frame; against ffmpeg the affected frames go from about
-4 dB SNR to 72 dB.

**machine/mpeg_demux: add an MPEG-1 system stream demultiplexer.**
Recovers one elementary stream and its timestamps, a byte at a time. Both
decoders on the board contain one and select their own stream by number.

**sound/gsc38gg307: add the CD-i full motion audio decoder.**
The Motorola GSC38GG307CF50 on GMPEG and VMPEG boards. Decoding is done with
`mpeg_audio`; the layer II framing is done in the device, since the frame
length covers padding and ancillary data the decoder does not read.

**video/mcd251: add the MPEG-1 full motion video decoder.**
The Motorola MCD251 on VMPEG boards, with the picture stores, the display
order, the display pacing and the picture DRAM it lets the host reach.
Decoding is done with `mpeg_video`. Only a technical summary is published,
but the registers it lists are at the offsets used here, so the register file
is the chip's own.

**cdi: add Digital Video Cartridge (VMPEG) emulation.**
The board that carries the two decoders and wires them to the player
(`src/mame/philips/cdidvc.cpp`). It is a card in a new slot, `dvc`, on
`cdimono1`, and is fitted by default: `-dvc ""` removes it. Dragon's Lair
then shows its own "requires a digital video board" screen, as on a player
without the cartridge.
Fix: https://github.com/mamedev/mame/issues/1170

**cdi: add NTSC support.**
The Mono-I is PAL: its slave answers the OS's video standard request (0xf6)
with PAL, and NTSC-only discs refuse to run. `cdimono1n` is the same player
as a separate clone set
Fix: https://github.com/mamedev/mame/issues/14018

## Notes on the review comments

- The bundled MPEG library is gone. Video decoding is @MagikalUnicorn's
  `mpeg_video` from #16053 and audio is MAME's `mpeg_audio`, so there is one
  MPEG decoder in the tree.
- The cartridge is no longer one device: `mcd251_device` and
  `gsc38gg307_device` are the chips, and `cdi_dvc_device` is the board around
  them, so an expansion card using either chip can reuse it.
- The chips are Motorola MCD251 (video) and GSC38GG307CF50 (audio), per
  https://github.com/cdifan/cdichips . The MCD270 is the integrated chip of
  the IMPEG boards, which is a different cartridge. The DSP56001 beside the
  audio decoder handles the attenuation ramp rather than the decoding; its
  microcode is downloaded from the cartridge ROM at boot, but MAME has no
  working DSP56000 core yet, so it is not emulated.
- The cartridge does not drive the speaker directly. On the real board its
  audio is a second pair of 16-bit channels, summed with the player's own in
  an analogue circuit, with an attenuator in front of each source: the pair of
  AD7528s the slave drives for the player's volume, and the ramp the
  cartridge's DSP56001 computes for the MPEG channels. So the decoder routes
  into the card, the card into the slot and the slot into the speaker, where
  the two are summed. Neither attenuator is emulated, and the driver does not
  model the player's volume control for its own audio either.
- The slot takes any cartridge now, not just this one. `device_cdi_dvc_interface`
  is what the player needs from a cartridge (its interrupt vector, vertical
  timing in, the video it produces), the slot carries the player's side (bus,
  interrupt, screen, video standard, audio) and `cdi_dvc_vmpeg_device` is the
  VMPEG board implementing it. An IMPEG cartridge, with its single MCD270,
  would be another option in the same slot and nothing in the driver would
  change. Note the cartridge ROM set is now `cdi_dvc_vmpeg` rather than
  `cdi_dvc`.
- The cartridge now maps itself into the SCC68070's space instead of the
  driver holding handlers for it, so none of its internals are in the machine
  description any more. What it does not map stays with the machine's
  catch-all bus error, which is also what a player with no cartridge fitted
  sees: Dragon's Lair still shows "Unsuitable Hardware" with -dvc "".
- The comments that explained the past, or the obvious, are gone.

## Tested (Both Manually + AI )

- Playing full motion video with audio intro && in-game without crash:
  7th Guest, Brain Dead 13, Christmas Country, Clue, Creature Shock, Drug Wars,
  Atlantis, Crime Patrol,  Earth Command, Effacer, Kingdom: The Far Reaches,
  The Last Bounty Hunter, Litil Divil, Lost Eden, The Lost Ride,
  Mad Dog McCree, Guignols de l'Info,  Mad Dog II: The Lost Gold, Shadoan,
  Thunder in Paradise,Caesars World of Boxing, Chaos Control, Christmas Crisis,
  Golden Oldies I, Jack Sprite vs. the Crimson Ghost, Lucky Luke, Space Ace,
  Tetsuo Gaiden, World Cup Golf, Monty Python's Invasion from the Planet Skyron.

- NTSC games with cdimono1n: Coneheads (USA) and The Adventure of the Space Ship
Beagle (Japan) play.

- The ten VMPEG test ROMs from the CDi_MiSTer simulator, ported to run on
  `cdimono1`, all pass: decoded pictures match ffmpeg at 48 dB PSNR or better,
  and the audio tone test matches its reference.

- Decoded audio was compared sample by sample against ffmpeg on streams taken
  from the discs, and matches.

- Splitting the cartridge into the two chips changed nothing: the audio of six
  titles and 118 snapshots from them are identical byte for byte before and
  after, as are the output and snapshots of all ten test ROMs.

## AI assistance disclosure

Developed with assistance from Claude Opus 5. The contributor reviewed the resulting changes and remains responsible for the contribution.

### CODE REVIEW

**Comment by mamehaze** [comment]
you may need to work with the @MagikalUnicorn ( https://github.com/mamedev/mame/pull/16053 ) as that PR is trying to establish a standard MPEG decoder for MAME, and it's likely only one decoder is going to be accepted.

**Comment by aderumier** [comment]
> you may need to work with the @MagikalUnicorn ( #16053 ) as that PR is trying to establish a standard MPEG decoder for MAME, and it's likely only one decoder is going to be accepted.

thanks, I'll look at it !

**Comment by angelosa** [comment]
1. This should also fix https://github.com/mamedev/mame/issues/1170 btw.
2. What MPEG chip the DVC really employs, isn't it one of the <https://github.com/cdifan/cdichips#mpeg-chips> so presumably this emulates a MCD251? Why DSP56k isn't handled then? 
3. If above is true, is it a fair assumption that the interaction needs encapsulating into a standard device so that tomorrow if an ISA16 or Amiga Zorro (at least) expansion card surfaces using this can be Lego-ized without extra work?
4. Is the slot option handled properly? What if tomorrow we want to emulate a non-VMPEG cart?

**Comment by MagikalUnicorn** [comment]
Thanks for the ping, probably worth paging @Vincent-Halver on CDi work if he hasn't seen this already too. Would be great to have NTSC support as I have a tonne of NTSC CDi Karaoke discs which need it.

I did consider using pl_mpeg for the Cobra3 work but chose the route of a first-principles implementation from the spec as it was one less library for MAME to depend on, Codex was able to do it with some effort, it mirrored the existing MAME mpeg_audio approach, and most importantly we'd have complete control over what internal decoder state we could expose to support a wide variety of hardware MPEG implementations, some of which I couldn't do with pl_mpeg without modifications to the pl_mpeg source. (Tangentially, there was also some kind of vulnerability Codex came across while working with pl_mpeg which sent it into a tailspin when I tried to work with it further, as it refused to disclose to me what it was, which was frustrating else I'd have at least opened an issue on the GitHub it to have it fixed).

I'm not part of the MAME team but I'd also suggest implementing the MPEG chip itself, so it can be reused elsewhere and the ownership of function is correctly scoped. That will likely mean you'll need deeper visibility of the decoder's internal picture reconstruction state. Could you look at that against my proposed mpeg_video decoder and see if there is any functionality or internal state exposure missing from it that would be needed to support your use case?

File: src/mame/philips/mcd212.cpp [hunk 1]
```diff
@@ -318,10 +318,17 @@ int mcd212_device::get_border_width()
 	return width;
 }
 
-uint32_t mcd212_device::get_backdrop_plane()
+uint32_t mcd212_device::get_backdrop_plane(int x, int y)
 {
 	if (BIT(m_image_coding_method, ICM_EV_BIT))
-		return 0; // External Video Background. Default to Black since there is no DVC.
+	{
+		// External Video Background: the DVC drives it when one is fitted and
```
Comment (Vincent-Halver) [inline src/mame/philips/mcd212.cpp]: Personally, I don't like Claude's verbose comments. Since the feature is implemented, I'd remove the comment as it is now self explanatory.

File: src/mame/philips/cdi.cpp [hunk 2]
```diff
@@ -354,13 +369,113 @@ void quizard_state::mcu_p3_w(uint8_t data)
...
+	{
+		bus_error_w((0xe80000 >> 1) + offset, data);
+		return;
+	}
+
+	if (m_dvc)
+		m_dvc->ram_w(offset, data, mem_mask);
+}
+
+/*************************
```
Comment (Vincent-Halver) [inline src/mame/philips/cdi.cpp:365]: I think some of the internals of the device are leaking into the machine description.

File: src/mame/philips/cdi.cpp [hunk 3]
```diff
@@ -532,10 +647,30 @@ void cdi_state::cdimono1(machine_config &config)
 	m_slave_hle->read_mousebtn().set_ioport("MOUSEBTN");
 	m_slave_hle->testplug_callback().set_ioport("TESTPLUG");
 
-	SOFTWARE_LIST(config, "cd_list").set_original("cdi").set_filter("!DVC");
+	// Digital Video Cartridge, fitted by default; -dvc "" removes it
+	CDI_DVC_SLOT(config, m_dvc_slot, cdi_dvc_cards, "vmpeg");
+	m_dvc_slot->set_option_machine_config("vmpeg", [this] (device_t *device) { dvc_config(device, true); });
+	m_dvc_slot->add_route(0, "speaker", 1.0, 0);
```
Comment (Vincent-Halver) [inline src/mame/philips/cdi.cpp:612]: I'm not sure about this, does the DVC directly drive the speaker? Or does it transfer audio data to the mixer?

File: src/mame/philips/cdi.cpp [hunk 4]
```diff
@@ -633,11 +768,10 @@ ROM_START( cdi490a )
 	ROM_SYSTEM_BIOS( 0, "cdi490", "CD-i 490" )
 	ROMX_LOAD( "cdi490a.rom", 0x000000, 0x80000, CRC(e2f200f6) SHA1(c9bf3c4c7e4fe5cbec3fe3fc993c77a4522ca547), ROM_BIOS(0) | ROM_GROUPWORD | ROM_REVERSE  )
 
-	ROM_REGION(0x60000, "mpegs", 0) // keep these somewhere
+	// The IMPEG variant of the DVC.  The VMPEG card's ROM now lives in the
```
Comment (Vincent-Halver) [inline src/mame/philips/cdi.cpp]: I think remove comment which explains the past (Claude loves to explain things that aren't true anymore in comments).

**Review by Vincent-Halver — CHANGES_REQUESTED** [review]

File: src/mame/philips/cdi.cpp [hunk 5]
```diff
@@ -731,6 +753,9 @@ ROM_START( cdimono1 )
 	ROM_LOAD( "zx405042p__cdi_slave_2.0__b43t__zzmk9213.mc68hc705c8a_withtestrom.7206", 0x0000, 0x2000, CRC(688cda63) SHA1(56d0acd7caad51c7de703247cd6d842b36173079) BAD_DUMP )
 ROM_END
 
+// same board and ROMs, see cdi_state::cdimono1n
+#define rom_cdimono1n rom_cdimono1
```
Comment (Vincent-Halver) [inline src/mame/philips/cdi.cpp:683]: Is this guard so it doesn't impact the other systems? That feels a bit weird to use #define in the middle of a file, and looks like it would be confusing.

**Review by Vincent-Halver — CHANGES_REQUESTED** [review]

**Comment by Vincent-Halver** [comment]
Commit 022789a21118526069ca697a83e8b35aba4f6a77 is the same fix from https://github.com/mamedev/mame/pull/15978 so that looks right to me. The PCB test there should still be pulled in, as those tests are able to show in lower level details possible errors and edge cases.

I think there should be a cdi_dvc device. Definitely agree that we'd want to avoid having a fork specific for CDI. I hadn't come to an answer on which decoder to use, and I'm not familiar enough to review the 3rd party library here. 

Mpeg1 audio has a decoder in MAME already. I was considering using that for audio.

It looks to me like the MPEG audio here is driving the speakers directly, but I'm not 100% sure that's right. Should it go to the audio mixer? 

For NTSC support, it looks pretty close to what I came up with. If it works with Beagle it seems right.

**Comment by Vincent-Halver** [comment]
>     2. What MPEG chip the DVC really employs, isn't it one of the https://github.com/cdifan/cdichips#mpeg-chips so presumably this emulates a MCD251? Why DSP56k isn't handled then?

I'm not an expert but I think DSP56K is correct. The mpeg roms are from later editions and so I think that's the chip they would be. In theory, the rom has all the decoder information and an external library isn't necessary at all.

File: src/mame/philips/cdi.cpp (hunk 5 above)
Comment (startaq) [reply src/mame/philips/cdi.cpp:683]: That's fine, it's a common pattern in MAME to duplicate ROM definitions.

**Comment by aderumier** [comment]
> 
> I'm not part of the MAME team but I'd also suggest implementing the MPEG chip itself, so it can be reused elsewhere and the ownership of function is correctly scoped. That will likely mean you'll need deeper visibility of the decoder's internal picture reconstruction state. Could you look at that against my proposed mpeg_video decoder and see if there is any functionality or internal state exposure missing from it that would be needed to support your use case?

yes, I'll look tomorrow. (I wasn't aware about your patches serie). cdi dvc model use same chip, so it seem to be a better way indeed. 
I have unit tests on my side, ported from mister fpga cdi core, so I could easily verify implementation
https://github.com/MiSTer-devel/CDi_MiSTer/tree/main/sim2/testroms

File: src/mame/philips/cdi.cpp (hunk 3 above)
Comment (aderumier) [reply src/mame/philips/cdi.cpp:612]: neither, exactly — the cartridge is a second audio channel that the mainboard sums with the player's own audio in an analogue circuit.

The evidence is in the FPGA core we used as the hardware reference. Its dual_ad7528_attenuation module models the real board's audio output stage and takes four inputs — CDIC left/right and VMPEG left/right — applies a volume factor to each, and sums them. Two things in it answer the question directly:

The attenuators are a pair of AD7528 dual DACs (7530 and 7550) on the mainboard, driven by the slave microcontroller's GPIOs through a 74HC164 shift register — that's the player's volume control, applied to both sources.

**Comment by aderumier** [comment]
> I'm not part of the MAME team but I'd also suggest implementing the MPEG chip itself, so it can be reused elsewhere and the ownership of function is correctly scoped. That will likely mean you'll need deeper visibility of the decoder's internal picture reconstruction state. Could you look at that against my proposed mpeg_video decoder and see if there is any functionality or internal state exposure missing from it that would be needed to support your use case?

I have updated the serie, reusing your mpeg_video decoder, why some additions for cdi here!

https://github.com/mamedev/mame/pull/16181/changes/ff0eb995955ee02e417b31d304ab9402aaad225a

**Comment by aderumier** [comment]
> 1. This should also fix [Philips CD-i: support for Digital Video Cartridge (DVC) #1170](https://github.com/mamedev/mame/issues/1170) btw.
> 2. What MPEG chip the DVC really employs, isn't it one of the https://github.com/cdifan/cdichips#mpeg-chips so presumably this emulates a MCD251? Why DSP56k isn't handled then?

Motorola MCD251 (video) and GSC38GG307CF50 (audio), both per cdifan's list, with the MCD251's technical summary confirming our register offsets. The DSP56001 does the attenuation ramp, not the decoding; its microcode is downloaded from the cartridge ROM at boot, but MAME's DSP56000 core is an empty skeleton (TODO: emulation), so it isn't emulated.


> 3. If above is true, is it a fair assumption that the interaction needs encapsulating into a standard device so that tomorrow if an ISA16 or Amiga Zorro (at least) expansion card surfaces using this can be Lego-ized without extra work?

Done — mcd251_device, gsc38gg307_device and mpeg_demux are standalone devices.

> 4. Is the slot option handled properly? What if tomorrow we want to emulate a non-VMPEG cart?

This one was not properly handled, and I've now fixed it. The slot took the concrete cartridge class, and the driver configured the card through a downcast lambda — so a second cartridge type couldn't plug in. Following MAME's waveblaster-connector pattern:

device_cdi_dvc_interface — what the player needs from any cartridge: its interrupt vector, vertical timing in, and the video it produces.
cdi_dvc_slot_device — carries the player's side: bus, interrupt, screen, video standard and audio. The driver configures the slot.
cdi_dvc_vmpeg_device — the VMPEG board implementing the interface, taking the host resources from its slot.

**Comment by angelosa** [comment]
> The DSP56001 does the attenuation ramp, not the decoding; its microcode is downloaded from the cartridge ROM at boot, but MAME's DSP56000 core is an empty skeleton (TODO: emulation), so it isn't emulated.

Uh, I missed the [memo](https://github.com/mamedev/mame/commit/450a57568e16822d625c3e0cd390e9c9e6a08718) nevermind this for now.

**Comment by MooglyGuy** [comment]
The above CI failures don't seem quite right, but please remove the unecessarily-verbose Claude comment as pointed out by @Vincent-Halver, aside from that, hopefully we can get both this change and his changes that this depends on in before EOW.

**Comment by aderumier** [comment]
> The above CI failures don't seem quite right, but please remove the unecessarily-verbose Claude comment as pointed out by @Vincent-Halver, aside from that, hopefully we can get both this change and his changes that this depends on in before EOW.

have rebase from last master && repush, as the CI failures was not mine.
I have remove  unecessarily-verbose Claude comment as much as possible, keeping the interesting comments, but if you really don't want any comments in code, just tell me.

**Comment by MooglyGuy** [comment]
You didn't mark the feedback points from @Vincent-Halver as resolved, so I misinterpreted the thread. This looks good to me at this point, hopefully the dependent pull requests can be merged soon.

File: src/mame/philips/cdi.cpp [hunk 6]
```diff
@@ -191,15 +192,28 @@ INPUT_PORTS_END
 *  Machine Initialization  *
 ***************************/
 
+void cdi_state::machine_start()
```
Comment (MooglyGuy) [inline src/mame/philips/cdi.cpp]: Note to self, remove this after merging.

File: src/mame/philips/cdi.cpp [hunk 7]
```diff
@@ -191,15 +192,28 @@ INPUT_PORTS_END
...
+	m_mcd212->set_ext_video_source(m_dvc_slot);
+
+	m_cdic_intreq = false;
+	m_dvc_intreq = false;
+	m_in4_owner = IN4_IDLE;
 }
 
 void quizard_state::machine_start()
 {
+	cdi_state::machine_start();
```
Comment (MooglyGuy) [inline src/mame/philips/cdi.cpp]: As above, remove irrelevant call to empty parent-class implementation.

File: src/mame/philips/cdi.cpp [hunk 8]
```diff
@@ -352,15 +366,67 @@ void quizard_state::mcu_p3_w(uint8_t data)
 *     DVC cartridge      *
 *************************/
 
-uint16_t cdi_state::dvc_r(offs_t offset, uint16_t mem_mask)
+/*************************
+*   IN4 IRQ arbitration  *
+*************************/
+
+// The CDIC and the DVC both drive IN4.  A small SR flip flop in the real
```
Comment (MooglyGuy) [inline src/mame/philips/cdi.cpp]: Decent comment, remove overly-pedantic double space between sentences.

**Comment by MooglyGuy** [comment]
According to Github there are conflicts that need resolving now that some dependent changes have been merged; if you feel up to it, you can address my small notes above, otherwise just resolve the merge conflicts and I'll take care of my feedback points post-merge.

**Comment by aderumier** [comment]
> According to Github there are conflicts that need resolving now that some dependent changes have been merged; if you feel up to it, you can address my small notes above, otherwise just resolve the merge conflicts and I'll take care of my feedback points post-merge.

I have merged conflict and retested to whole games .  
I'll dont have too much time until next week, so feel free to fix what you want post-merge.

(and thanks again to everyone for the reviews ! )
